src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -768,14 +770,35 @@ pub enum StaleItem {
 impl LocalFingerprint {
     /// Read the environment variable of the given env `key`, and creates a new
     /// [`LocalFingerprint::RerunIfEnvChanged`] for it.
-    ///
-    // TODO: This is allowed at this moment. Should figure out if it makes
-    // sense if permitting to read env from the config system.
     #[allow(clippy::disallowed_methods)]
-    fn from_env<K: AsRef<str>>(key: K) -> LocalFingerprint {
-        let key = key.as_ref();
+    fn from_env<K: AsRef<str>>(
+        key: K,
+        envs: &BTreeMap<String, Option<OsString>>,
+    ) -> LocalFingerprint {
+        fn get_envs_case_insensitive(
+            key: &str,
+            envs: &BTreeMap<String, Option<OsString>>,
+        ) -> Option<OsString> {
+            let upper_case_key: String = key.to_uppercase();
+            for (k, v) in envs {
```

Comment:
This conversion seems to be in a deeply nested loop. M `rerun_if_env_changed` * N `[env]` keys on Windows. We should probably find a way to do it once and for all.

This doesn't really seem to be a performance issue for general usage though.

Reply:
A small fix would be to check the map with the key as-is, and then only check for the case-insensitive if that fails. 99% of the time the case will match. [`EnvConfig::get_env_os`](https://github.com/rust-lang/cargo/blob/a7917fdc3ae499b2324e45b94ed7fd08cc242b1c/src/cargo/util/context/environment.rs#L115-L126) does something similar.